src/protocol/mock.rs

```diff
@@ -12,6 +12,126 @@ impl MockProtocol {
+                let command_id = 0x24;   // Command Id for Control Byte to write blocks to transponder's RF blocks
+                let mode = 0x01; // addressed
+                let db_n = 0x01;
+                let db_size = 0x04;
+                let addr = 0x00;
+                
+                let mut data: smallvec::SmallVec<[u8; 32]> = smallvec::smallvec![command_id, mode, 0, 1, 2, 3, 4, 5, 6, 7, addr, db_n, db_size];
+                let mut num_bytes = 3;
+                let mut cmd_op;
+                let mut act_cnt8 = 5;
+                let mut is_actuators = false;
+                let mut mem_blk1 = vec![0,0];
+                let mut mem_blk2:Vec<u8> = vec![];
+                let mut mem_blk3:Vec<u8> = vec![];
+
+                let command;
+                if op_mode_block.is_some(){
+                    let bl = op_mode_block.as_ref().unwrap();
+                    log::debug!("Num act blks: {:#?}, cmd_op: {:#?}, command: {:#?} ",bl.act_cnt8,bl.cmd_op,bl.command);
+                    command = bl.command;
+                    cmd_op = bl.cmd_op;
+                    if command != 0 {
+                        //Not all off command
+                        if actuator_mode_blocks.is_some() && cmd_op != 0 {
+                            let mut last_int = 0;
+                            let bl = actuator_mode_blocks.as_ref().unwrap();
+                            let mut blks = vec![];
+                            if bl.block0_31.is_some(){ 
+                                mem_blk1.append(&mut vec![0,bl.block0_31.as_ref().unwrap().b0]); 
+                                blks.extend([bl.block0_31.as_ref().unwrap().b1, bl.block0_31.as_ref().unwrap().b2, bl.block0_31.as_ref().unwrap().b3].iter().copied());
+                            }
+                            if bl.block32_63.is_some(){ blks.extend([bl.block32_63.as_ref().unwrap().b0, bl.block32_63.as_ref().unwrap().b1, bl.block32_63.as_ref().unwrap().b2, bl.block32_63.as_ref().unwrap().b3].iter().copied());}
+                            if bl.block64_95.is_some(){ blks.extend([bl.block64_95.as_ref().unwrap().b0, bl.block64_95.as_ref().unwrap().b1, bl.block64_95.as_ref().unwrap().b2, bl.block64_95.as_ref().unwrap().b3].iter().copied());}
+                            if bl.block96_127.is_some(){ blks.extend([bl.block96_127.as_ref().unwrap().b0, bl.block96_127.as_ref().unwrap().b1, bl.block96_127.as_ref().unwrap().b2, bl.block96_127.as_ref().unwrap().b3].iter().copied());}
+                            if blks.len() != 0 {
+                                is_actuators = true;
+                                let mut cnt = 0;
+                                for blk in blks.iter_mut() { //find last relavant byte
+                                    cnt += 1;
+                                    if *blk != 0 { last_int = cnt; }
+                                }
+                                blks.truncate(last_int); //remove unneeded bytes (trailing zeros)
```

Comment:
This function is full of code that isn't using Rust's zero cost abstractions. They are very readable and sometimes faster. Intermittently, read code samples in the docs as you go to see how they recommend using collections, iterators, etc.

Example alternative: https://godbolt.org/z/Tx9f5h
A short blog: https://carette.xyz/posts/zero_cost_abstraction/

Usage of standard library features translates well to other frameworks you may encounter in your career. You can reference the design concepts and hone your skills the more experience you get.